File: O/source_code_4.py
def warshall_closure(matrix):
    n = len(matrix)
    for k in range(n):
        for i in range(n):
            for j in range(n):
                matrix[i][j] = matrix[i][j] or (matrix[i][k] and matrix[k][j])
    return matrix

# Input
n = int(input("Enter the number of vertices: "))
path = []

print("Enter the adjacency matrix:")
for _ in range(n):
    row = list(map(int, input().split()))
    path.append(row)

# Compute transitive closure
closure = warshall_closure(path)

# Output
print("\nTransitive Closure Matrix:")
for row in closure:
    print(" ".join(map(str, row)))

File: O/source_code_4.cpp
#include <iostream>
using namespace std;

int main()
{
    int n;
    cout << "Enter the number of vertices: ";
    cin >> n;
    int path[10][10];

    cout << "Enter the adjacency matrix:\n";
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            cin >> path[i][j];
        }
    }

    for (int k = 0; k < n; k++)
    {
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < n; j++)
            {
                path[i][j] = path[i][j] || (path[i][k] && path[k][j]);
            }
        }
    }

    cout << "\nTransitive Closure Matrix:\n";
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            cout << path[i][j] << " ";
        }
        cout << endl;
    }
    return 0;
}